raw-spans-grouper/raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansProcessor.java

```diff
@@ -196,7 +242,101 @@ public void process(Record<TraceIdentity, RawSpan> record) {
         .record(Duration.between(start, Instant.now()).toMillis(), TimeUnit.MILLISECONDS);
     // no need to do context.forward. the punctuator will emit the trace once it's eligible to be
     // emitted
-    return;
+  }
+
+  private void processSpanForPeerServiceNameIdentification(
+      TraceIdentity key, RawSpan value, TraceState traceState, long currentTimeMs) {
+    Event event = value.getEvent();
+    String tenantId = key.getTenantId();
+    ByteBuffer traceId = value.getTraceId();
+    boolean firstEntry = (traceState == null);
+    Optional<String> maybeEnvironment = HttpSemanticConventionUtils.getEnvironmentForSpan(event);
+    if (SpanSemanticConventionUtils.isClientSpanForOCFormat(
+        event.getAttributes().getAttributeMap())) {
+      handleClientSpan(tenantId, event, maybeEnvironment.orElse(null));
+    } else {
+      handleServerSpan(tenantId, event, maybeEnvironment.orElse(null));
+    }
+
+    Timestamps timestamps =
+        traceLatencyMeter.trackEndToEndLatencyTimestamps(
+            currentTimeMs, firstEntry ? currentTimeMs : traceState.getTraceStartTimestamp());
+    StructuredTrace trace =
```

Comment:
We build structureTrace in two places - 1) TraceEmitPuncutator 2) Bypass condition.

This seems a third place. What is the condition for this - it's not very clear. Will the same trace be also emitted via TraceEmitPuncuator.

Reply:
No, the same trace be also emitted via TraceEmitPuncuator as we return in the caller function: https://github.com/hypertrace/hypertrace-ingester/blob/aed68f227e13e6914c295a2f789818410e454f08/raw-spans-grouper/raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansProcessor.java#L202

Reply:
Added a comment